src/map/Pins.js

```diff
@@ -28,49 +29,127 @@
     };
 
+    Pins.prototype.pinPathCoords = function (shape) {
+        var _path = '';
+        switch(shape){
+            case "circle":
+                _path += "m-24,-31c0,10.8335 6.8925,20.0592 16.53,23.53l8.47,8.47l8.47,-8.47c9.6375,-3.4708 16.53,-12.6965 16.53,-23.53c0,-13.8073 -11.1927,-25 -25,-25c-13.8073,0 -25,11.1927 -25,25z";
+                break;
+            case "pin":
+                _path += "M 0,0 C -2,-20 -10,-22 -10,-30 A 10,10 0 1,1 10,-30 C 10,-22 2,-20 0,0 z M -2,-30";
+                break;
+            case "square":
+                var _squareWidth = 50;
+                var _cornerRadius = 6;
+                var _pointHeight = 10;
+                var _pointWidth = 20;
+                _path += 'M0 0';
+                _path += 'L-'+(_pointWidth/2)+' -'+(_pointHeight);
+                _path += 'L-'+((_squareWidth/2)-_cornerRadius)+' -'+(_pointHeight);
+                _path += 'Q-'+(_squareWidth/2)+' -'+(_pointHeight)+' -'+(_squareWidth/2)+' -'+(_pointHeight+_cornerRadius);
+                _path += 'L-'+(_squareWidth/2)+' -'+((_pointHeight+_squareWidth-_cornerRadius));
+                _path += 'Q-'+(_squareWidth/2)+' -'+((_pointHeight+_squareWidth))+' -'+(_squareWidth/2-_cornerRadius)+' -'+((_pointHeight+_squareWidth));
+                _path += 'L'+((_squareWidth/2)-_cornerRadius)+' -'+((_pointHeight+_squareWidth));
+                _path += 'Q'+(_squareWidth/2)+' -'+((_pointHeight+_squareWidth))+' '+(_squareWidth/2)+' -'+((_pointHeight+_squareWidth-_cornerRadius));
+                _path += 'L'+(_squareWidth/2)+' -'+(_pointHeight+_cornerRadius);
+                _path += 'Q'+(_squareWidth/2)+' -'+(_pointHeight)+' '+((_squareWidth/2)-_cornerRadius)+' -'+(_pointHeight);
+                _path += 'L'+(_pointWidth/2)+' -'+(_pointHeight);
+                _path += 'Z';
+                break;
+        }
+        return _path;
+    };
     Pins.prototype.layerZoomed = function (base) {
         Layer.prototype.layerZoomed.apply(this, arguments);
         this.pinsPaths
-            .attr("transform", function (d) {
+            .style("display", function (d) {
+                var pos = base.project(d[0], d[1]);
+                if (!pos) {
+                    return "none";
+                }
+                return null;
+            })
+            .attr("transform", function (d,i) {
                 var pos = base.project(d[0], d[1]);
                 if (!pos) {
                     pos = [0, 0];
                 }
-                return "translate(" + pos[0] + ", " + pos[1] + ")scale(" + 0.5 + ")";
+                d3.select(".pin-icon-"+i).attr("transform","translate("+pos[0]+", "+pos[1]+")scale("+0.5+")");
+                d3.select(".pin-text-"+i).attr("transform","translate("+pos[0]+", "+pos[1]+")scale("+0.5+")");
+                return "translate(" + pos[0] + ", " + pos[1] + ")scale(" + 1 /* * base._zoom.scale() */ + ")";
             })
+            .attr("stroke-width", 1 + "px")
```

Comment:
FYI - Anything inside "layerZoomed" is very performance sensitive - in general we get away with simply applying a transform on the outer layer (for geohash and choropleths), but pins need to keep their size no matter the zoom level, hence each one gets the transformation.